### DISCUSSION TITLE
adding a password meter

### DISCUSSION THREAD

**Discussion by badhri17** [formkit/formkit#830] [Ideas]
I have an idea for adding a strength meter displayed as a popup while a value is being entered.

I am ready to contribute to the project by adding this feature  

---

## example 

![chrome-capture-2023-6-13](https://github.com/formkit/formkit/assets/60823745/4d969b57-f839-481b-8546-5e683c7fd650)

## example 2

![image](https://github.com/formkit/formkit/assets/60823745/5abad30d-03e5-4884-b00d-6e83a13e2510)

**Comment by fenilli** [comment]
This feature could be added using a plugin in the addons, but there are some questions about it, how would it know what is weak and strong?

**Comment by badhri17** [comment]
thank you gustavo for your response 
first can you share with the plugin you talk about because I have checked the docs and nothing similar there

and about your question I am thinking of it as a props

there is 3 states: strong, medium,weak so: mediumRegex,strongRegex
and for the labels (the messages): weakLabel,strongLabel,mediumLabel

we need these properties to achieve this functionality perfectly

**Comment by fenilli** [comment]
I didn't mean that we had this already, I meant that this feature shouldn't be in the core of the input itself as there are people who would not opt-in for it, and its not a necessary feature.

I meant that you could write an addon/plugin to achieve this functionallity and pass a PR to the addons package.

It would add 6 props, I think that is too much, better to have instead a single prop that receiveis an object with those, also how would it integrate with validation? also if someone wants a simpler version that would be aside from the help instead of a popup how would that be possible if so?

**Comment by badhri17** [comment]
sorry for the misunderstanding and yes I agree it will be better as an addon 
for the implementation I am thinking of this way:

```vue
  <FormKit
    type="password"
    label="enter a passowrd"
    :passwordMeter="{
    'strongRegex':'^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[@#$%^*!~]).{12,}$',
    'mediumRegex':'^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[@#$%^*]).{8,}$',
    'weakLabel':'weak',
    'mediumLabel': 'medium',
    'strongLabel' : ' strong'
    }"
  />
``` 

by default the result :

![chrome-capture-2023-6-14](https://github.com/formkit/formkit/assets/60823745/ed9f3972-69e9-4fb2-89f0-68b8890c8295)



and they can custom the popup with slots

```vue
  <FormKit
    type="password"
    label="enter a passowrd"
    :passwordMeter="{
    'strongRegex':'^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[@#$%^*!~]).{12,}$',
    'mediumRegex':'^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[@#$%^*]).{8,}$',
    'weakLabel':'password is weak',
    'mediumLabel': 'password is good',
    'strongLabel' : 'password is very strong'
    }"
  >
  <template #header>
        <h6>Pick a password</h6>
    </template>
    <template #footer>
        <Divider />
        <p class="mt-2">Suggestions</p>
        <ul class="pl-2 ml-2 mt-0" style="line-height: 1.5">
            <li>At least one lowercase</li>
            <li>At least one uppercase</li>
            <li>At least one numeric</li>
            <li>Minimum 8 characters</li>
        </ul>
    </template>
  </formkit>
```
and the result 

<img width="315" alt="image" src="https://github.com/formkit/formkit/assets/60823745/78ca9e50-0faa-4f8e-833b-8b22e00b883b">


and maybe we can add later other options if the user want the password meter to appear below the field not on a popup 


for the validation since its only a password meter plugin we don't have to add this option user can validate normally with the validation props

**Comment by fenilli** [comment]
Would those be optional or not? you sent this example, as the weak has not regex, by default would it be mean any value outside of empty is weak? if so, what would happen if there isn't a `mediumRegex`, but weak and strong?
```js
{
    'strongRegex':'^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[@#$%^*!~]).{12,}$',
    'mediumRegex':'^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[@#$%^*]).{8,}$',
    'weakLabel':'password is weak',
    'mediumLabel': 'password is good',
    'strongLabel' : 'password is very strong'
}
```

@justin-schroeder any input you could give?

**Reply by badhri17** [reply]
yes it will be optional user can put passwordMeter as a prop and the plugin will work automatically no need for the object

```vue
<FormKit
    type="password"
    label="enter a passowrd"
    passwordMeter
  />
```
by default the plugin has regex and labels

we only need two regex for the medium and strong if nothing match one of those it will classify as a weak password

what is your opinion?

**Reply by fenilli** [reply]
It should be fine as a addon yes, if you want you can work to make a PR for it.

**Reply by badhri17** [reply]
ok I will start working on it from today then I will make a PR for it
